lighthouse-core/audits/uses-rel-preconnect.js

```diff
@@ -0,0 +1,114 @@
+/**
+ * @license Copyright 2018 Google Inc. All Rights Reserved.
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License. You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
+ */
+
+'use strict';
+
+const Audit = require('./audit');
+const Util = require('../report/v2/renderer/util');
+const UnusedBytes = require('./byte-efficiency/byte-efficiency-audit');
+const URL = require('../lib/url-shim');
+const PRECONNECT_SOCKET_MAX_IDLE = 10;
+
+const learnMoreUrl =
+  'https://developers.google.com/web/fundamentals/performance/resource-prioritization#preconnect';
+
+class UsesRelPreconnectAudit extends Audit {
+  /**
+   * @return {!AuditMeta}
+   */
+  static get meta() {
+    return {
+      category: 'Performance',
+      name: 'uses-rel-preconnect',
+      description: 'Avoid multiple, costly round trips to any origin',
+      informative: true,
+      helpText: 'Consider using<link rel="preconnect dns-prefetch"> to set up early connections ' +
+        ' before an HTTP request is actually sent to the server. This will reduce multiple, ' +
+        `costly round trips to any origin. [Learn more](${learnMoreUrl}).`,
+      requiredArtifacts: ['devtoolsLogs'],
+      scoringMode: Audit.SCORING_MODES.NUMERIC,
+    };
+  }
+
+  /**
+   * Check is the connection is already open
+   * @param {!WebInspector.NetworkRequest} record
+   * @return {!boolean}
+   */
+  static hasAlreadyConnectedToOrigin(record) {
+    return record.timing && record.timing.dnsEnd - record.timing.dnsStart === 0 &&
+      record.timing.connectEnd - record.timing.connectStart === 0;
+  }
+
+  /**
+   * @param {!Artifacts} artifacts
+   * @return {!AuditResult}
+   */
+  static async audit(artifacts) {
+    const devtoolsLogs = artifacts.devtoolsLogs[UsesRelPreconnectAudit.DEFAULT_PASS];
+    let maxWasted = 0;
+
+    const [networkRecords, mainResource] = await Promise.all([
+      artifacts.requestNetworkRecords(devtoolsLogs),
+      artifacts.requestMainResource(devtoolsLogs),
+    ]);
+
+    const preconnectResults = {};
+    networkRecords
+      // filter out all resources that have a different origin
+      .filter(record => !URL.originsMatch(mainResource.url, record.url))
+      // filter out all resources that are not loaded by the document
+      .filter(record => {
+        return record.initiatorRequest() !== mainResource && record !== mainResource;
+      // filter out urls that do not have an origin
+      }).filter(record => {
+        return !!URL.getOrigin(record.url);
+      // filter out all resources where origins are already resolved
```

Comment:
since we're only displaying origins and not full request URLs, I feel like this is already taken care of by reduction down to unique origins below.

(The assumption being that connect is always a cost for new origins... which seems a reasonable assumption. right @patrickhulce ?)

Reply:
yeah but he's trying to find the first representative network record for the origin so you do need to filter out the network records that were already connected